### ISSUE TITLE
Vision for "Split" pipeline

### ISSUE THREAD

**Issue by Iftahh** [WeAreAigur/client#15]
Split pipelines will enable parallelism - there are two different ways I see to define such a pipeline  - manual and automatic

Use case - the user enters a song summary and then the pipeline generates song lyrics and album cover at the same time.

---

1. "automatic" solution - 

The execution logic will find out which nodes can be run in parallel and handle it automatically.
In the example below the execution logic will discover that node-2 does not depend on other nodes, so it can run in parallel to node-0,   and afterwards  node-3 can run in parallel to node-1.   Node-5 depends on both node-4 and 1 so it will automatically wait for both of them to complete.

*The upside*:   easy for the developer - nothing special to set up
*The downside*:  the input/output must be explicitly and completely be defined. If the `action` of node-x implicitly depends on the output of earlier nodes, without defining it as part of the input then the execution logic will mess up and activate it before the right time.
Another downside - this solution is quite complex to implement.  Building the dependency graph and managing the order of nodes to run is one problem, and detecting which node's output is accessed in the input is another challenge.

```javascript
            flow
  /* 0 */   .node(replaceString, ({ input }) => ({
                text: input.songDesc,
                modifier: 'The lyrics of a song about $(text)$',
            }))
  /* 1 */   .node(gpt3Prediction, ({ prev }) => ({
                prompt: prev.text,
            }))
  /* 2 */  .node(replaceString, ({ input }) => ({
                text: input.songDesc,
                modifier: `artistic, album cover, $(text)$, highly detailed, cinematography`,
            }))
  /* 3 */   .node(stabilityTextToImage, ({ prev }) => ({
                text_prompts: [
                    { text: prev.text, },
                ],
                steps: 60,
            }))
  /* 4 */  .node(supabaseUpload, ({ prev }) => ({
                bucket: 'results',
                extension: 'png',
                file: prev.result,
                supabaseServiceKey: process.env.SUPABASE_SERVICE_KEY!,
                supabaseUrl: process.env.SUPABASE_URL!,
            }))
  /* 5 */  .output(({ prev, nodes }) => ({
                url: nodes[4].output.url,
                lyrics: nodes[1].output.text,
            })),

```

-------

2. "manual" solution - 
The developer sets up the pipeline in a way to explicitly tells the execution which nodes should be run in parallel - the above use case could look something like below  (this is off the top of my head, perhaps we can find a cleaner API)

*The upside*:  easy to implement 
*The downside*:   require manual decision to split the pipeline,  possibility for the programmer to miss optimization potential

```javascript
            flow
              .node(splitPipeline, (split) => {
                    split.branch().
                    /* 0 */   .node(replaceString, ({ input }) => ({
                                  text: input.songDesc,
                                  modifier: 'The lyrics of a song about $(text)$',
                              }))
                    /* 1 */   .node(gpt3Prediction, ({ prev }) => ({
                                  prompt: prev.text,
                              }))
                    split.branch()
                      /* 2 */  .node(replaceString, ({ input }) => ({  
     // accessing `prev` here would be the prev to the `splitPipeline` node!  It is not possible to access output of node-1 here.
                                    text: input.songDesc,
                                    modifier: `artistic, album cover, $(text)$, highly detailed, cinematography`,
                                }))
                      /* 3 */   .node(stabilityTextToImage, ({ prev }) => ({
                                    text_prompts: [
                                        { text: prev.text, },
                                    ],
                                    steps: 60,
                                }))
                      /* 4 */  .node(supabaseUpload, ({ prev }) => ({
                                    bucket: 'results',
                                    extension: 'png',
                                    file: prev.result,
                                    supabaseServiceKey: process.env.SUPABASE_SERVICE_KEY!,
                                    supabaseUrl: process.env.SUPABASE_URL!,
                                }))
           })  // end of splitPipeline       
  /* 5 */  .output(({ prev, nodes }) => ({
                url: nodes[4].output.url,
                lyrics: nodes[1].output.text,
            })),

```

**Comment by yairhaimo** [comment]
Automatic optimization sounds nice but I think it's a bit premature. Let's keep the API straightforward even if its a bit more verbose. We can always add an `.auto` builder method or `optimize` configuration flag later.

Regarding the manual definition maybe we should use a dedicated `.split` method (parallel to `.node`):

```ts
flow
  .split(
    flow.node(...).node(...),
    flow.node(...).node(...)
  )
```
More things to consider:
- Take into consideration conditional splits when implementing this. Conditional splits can be implemented later but lets think about it so we won't break an API or have a convoluted one just not to break it.
- When accessing `prev` after `split`, what does it contain? 
  - a merge of all the last nodes of each branch? what about properties with the same name?
    ```ts
    flow
      .split(
        flow.node(...).node(...), // <-- output is {text: string, name: string}
        flow.node(...).node(...) // <-- output is {name: string, age: number}
      )
     .node(({prev}) => // what is "prev.name"?
    ```
  - prev will be an array? `prev[0].name` `prev[1].name`. I think this is the most straightforward choice but good luck to us with the typing 😅

**Comment by yairhaimo** [comment]
TL;WR - not relevant atm
---

We can have `split` and `conditionalSplit` be the same "animal" if we let `prev`/`input`/`nodes` be accessed directly from the `flow` builder:

```ts
flow
  .node(blabla, () => ({}))
  .split(
      flow.prev.text === 'yes'
        ? flow.node(...)
        : null,
      flow.input.subject === 'cars'
        ? flow.node(...)
        : null
  )
  .output(() => ...)
```

Edit: On the other hand, it will pollute the `flow` api for most cases

Edit2: Unlesssss we change the API and drop the "input setter function" in favor of a plain object:

Before:
```ts
  flow
    .node(replaceString, ({ input }) => ({
      text: input.subject,
      modifier: 'tell me a joke about $(text)$',
    }))
    .node(gpt3Prediction, ({ prev }) => ({
      prompt: prev.text,
    }))
    .output(({ prev }) => ({
      joke: prev.text,
    })),
```

After:
```ts
  flow
    .node(replaceString, {
      text: flow.input.subject,
      modifier: 'tell me a joke about $(text)$',
    })
    .node(gpt3Prediction, {
      prompt: flow.prev.text,
    })
    .output({
      joke: flow.prev.text,
    }),
```

I need to make sure that it's feasible and that I'm not forgetting anything.


Edit 3 😅 - it breaks the typings (obviously in retrospect), back to the drawing board

**Comment by yairhaimo** [comment]
To continue my last futile attempt, we can have the `split` function receive `input`/`nodes`/`prev` as well and that should work:

Regular split:
```ts
flow
  .node(blabla, () => ({}))
  .split(
      flow => flow.node(...),
      flow => flow.node(...),
  )
  .output(() => ...)
```

Conditional split
```ts
flow
  .node(blabla, () => ({}))
  .split(
      (flow, { prev }) => prev.text === 'yes'
        ? flow.node(...)
        : null,
      (flow, { input }) => input.subject === 'cars'
        ? flow.node(...)
        : null
  )
  .output(() => ...)
```

**Comment by Iftahh** [comment]
after a long break, I'm back to working on this feature.
It's quite mind bending to get the types correctly!  (my typescript-fu is still weak, but this is excellent training)

Regarding your first suggestion - 
I believe we have to pass a parameter function to `split` that would get the current flowBuilder instance as parameter, or else we can't access the correct `prev`, and we can't build the correct type for the new flow builder(s).
I propose the name for this function parameter: `splitter`

eg.  
```ts
 flow.node(...).node(..)
   .split((splitBuilder) => [
     splitBuilder.node(...).node(...),
     splitBuilder.node(...).node(...)
   ])
 ```
 
^ this is what I started implementing.

> More things to consider:
> 
> * Take into consideration conditional splits when implementing this. Conditional splits can be implemented later but lets think about it so we won't break an API or have a convoluted one just not to break it.

I think this works fine with a single splitter function, but then the function will have to have to get the runtime `prev` output result and `input` (not only the prev flow builder as above).  See example below.

However, I think *typing* the pipeline nodes for conditional nodes (ie. when it is changing in runtime) is nearly impossible.

I'm going to postpone the conditionals for later, and keep the splitter function with "simple" compile-time input and output  -  it will have only a single parameter - the previous FlowBuilder instance (known at compile time) and the output (ie. like the above example)

Example of how I think conditional split can work:
eg.   choose between one branches A  if the prev output is 'yes',  otherwise choose two parallel branches B.

```ts
 flow.node(...).node(..)
   .split((splitBuilder, {prev, input}) =>  
         prev.text === 'yes' ? [
             splitBuilder.node(...).node(...)   // A
         ] : [ 
             splitBuilder.node(...).node(...),  // B
             splitBuilder.node(...).node(...)
         ]
     )
)
 ```


> * When accessing `prev` after `split`, what does it contain?
>   
>   * a merge of all the last nodes of each branch? what about properties with the same name?
>    ...
>   * prev will be an array? `prev[0].name` `prev[1].name`. I think this is the most straightforward choice but good luck to us with the typing 😅

I'm trying to implement the 2nd choice - prev will resolve to array of the outputs of the last nodes (ie. the last of each split branch).

----
Another complication of split -

For the purpose of `node:start` and end events,  I think `index` is not enough, because once there are multiple branches the index won't tell you in which branch the node is. So, in addition to `index` (or completely replace it?) I'll pass an `id` which will be hierarchical, something like x.y.z.  
I'm pondering how to differentiate between nodes and branches. Currently I'm thinking of dot separated path with a `b` prefix for index of branches ie. `<index>.b<index>.<index>`

eg. in the first example above:

```ts
 flow.node(..A.).node(.B.)
   .split((splitBuilder) => [
     splitBuilder.node(..C.).node(..D.),
     splitBuilder.node(..E.).node(..F.)
   ])
 ```

when the `node:start` event is called  the `id` for nodes will be:
```
A: "0"
B: "1"
C: "2.b0.0"
D: "2.b0.1"
E: "2.b1.0"
F: "2.b1.1"
```

A few thoughts about this-
* if we want to use the id, for example, to access the output of earlier nodes, it would be nice to have a prefix `n` for nodes, so we can access it with recursive objects notation  instead of string keys,  eg.   `context.n2.b1.n0.output`  instead of `context.outputs["2.b1.0"]`
* conditional splits make typing this super complex. 
* Note the `split` actually generates a node behind the scenes (index 2 above).  Should this node also call `node:start` and `end` ?   maybe there should be a new event  `split:start` ?    and what about each branch,  should it have `branch:start` and end, like there is for the pipeline?
